doc/sphinx/src/mesh/mesh.rst

```diff
@@ -50,3 +50,64 @@ member.
   time-integration advance. The default behavior calls to each package's
   (StateDesrcriptor's) ``PreStepDiagnostics`` method which, in turn,
   delegates to a ``std::function`` member that defaults to a no-op.
+
+Multi-grid Grids Stored in ``Mesh``
+-----------------------------------
+
+If the parameter ``parthenon/mesh/multigrid`` is set to ``true``, the ``Mesh``
+constructor and AMR routines populate both 
+``std::vector<LogicalLocMap_t> Mesh::gmg_grid_locs`` and 
+``std::vector<BlockList_t> gmg_block_lists``, where each entry into the vectors 
+describes one level of the of the geometric multi-grid (GMG) mesh. For refined 
+meshes, each GMG level only includes blocks that are at a given logical level 
+(starting from the finest logical level on the grid and including both internal 
+and leaf nodes in the refinement tree) as well as leaf blocks on the next coarser 
+level that are neighbors to finer blocks, which implies that below the root grid 
+level the blocks may not cover the entire mesh. For levels above the root grid, 
+blocks may change shape so that they only cover the domain of the root grid. Note 
+that leaf blocks may be contained in multiple blocklists, and the lists all point
+to the same block (not a separate copy).
+
+*GMG Implementation Note:*
+The reason for including two levels in the GMG block lists is for dealing with 
+accurately setting the boundary values of the fine blocks. Convergence can be poor 
+or non-exististent if the fine-coarse boundaries of a given level are not 
+self-consistently updated (since the boundary prolongation from the coarse grid to 
+the fine grid also depends on interior values of the fine grid that are being updated 
+by a smoothing operation). This means that each smoothing step, boundary communication 
+must occur between all blocks corresponding to all internal and leaf nodes at a given 
+level in the tree and with all leaf nodes at the next coarser level which abut blocks 
+at the current level. Therefore, the GMG block lists contain blocks at two levels, but 
+smoothing operations etc. should usually only occur on the subset of those blocks that 
+are at the fine level.
```

Comment:
It's a bit in the weeds, but essentially the multigrid levels need to include all leaf and internal blocks at a given refinement level and all of their leaf block neighbors at the next coarser level. Operations such as smoothing are not performed on the coarser leaf blocks, but they have to be there so that the boundaries of *all* the fine blocks are set consistently during boundary communication.

Reply:
Right---I guess this is the part I don't understand. At what cadence are you setting the boundaries of the fine level? Every smoothing operation? And this is in contrast to what? What was the previous pattern that you tried that didn't work?

Reply:
Originally, I was setting the boundary values before smoothing and holding them fixed (and the way I wrote it we only the gmg levels to have all internal and leaf blocks at a given level). This converged for pure MG, but didn't work as a preconditioner for BiCGStab. Of course, this mean that the boundaries are not completely consistent during subsequent smoothing operations since prolongation into coarse-fine boundaries has a stencil that reaches into the fine interior zones.

Reply:
Did you only set them once per V-cycle? My main point of confusion is what's different between the new working algorithm and the old one.

Reply:
IIRC the way it worked was that they were set at the beginning of the way up the v-cycle and on the beginning of the way down at each level (on the way down they were set by prolongating from the next coarser gmg grid). This meant that at the beginning of smoothing the blocks at a given refinement level that had neighbors at a coarser level had those coarse-fine boundaries set once (so they looked like a constant Dirichlet boundary condition). Further boundary communication that occurs during smoothing iterations did not change the values of these coarse-fine boundaries. 

Now, with the two-level composite grid, every communication step during smoothing updates the values of these coarse-fine boundaries. The active values on the coarse blocks next to the boundaries are unchanged by the smoothing on fine blocks (and these get repeatedly communicated to the coarse buffer of a neighboring fine block), but the values in the interior of the fine block coarse buffer are changed by the smoothing iterations.

Reply:
I see---this makes sense. And the need for this is because the prolongation stencil requires information in the fine block interior right? Thanks for the explanation!

Reply:
> And the need for this is because the prolongation stencil requires information in the fine block interior right?

Yes, exactly. In retrospect it seems obvious, but figuring this out was a painful process for me...